codalab/server/bundle_manager.py

```diff
@@ -363,26 +363,42 @@ def _schedule_run_bundles_on_workers(self, workers, staged_bundles_to_run, user_
         # Build a dictionary which maps from uuid to running bundle and bundle_resources
         running_bundles_info = self._get_running_bundles_info(workers, staged_bundles_to_run)
 
+        # We pre-compute the workers available to each user (and the codalab-owned workers),
+        # such that workers that come online or regain the necessary resources while we
+        # are attempting to run each staged bundle will respect the ordering of
+        # staged_bundles_to_run (i.e., they won't be used immediately, and will be instead
+        # assigned bundles on the next run of _run_iteration).
+        resource_deducted_user_owned_workers = {}
+        user_parallel_run_quota_left = {}
+        for user in user_queue_positions.keys():
+            resource_deducted_user_owned_workers[user] = self._deduct_worker_resources(
+                workers.user_owned_workers(user), running_bundles_info
+            )
+            user_parallel_run_quota_left[user] = self._model.get_user_parallel_run_quota_left(
+                user, user_info_cache[user]
+            )
+        resource_deducted_codalab_owned_workers = self._deduct_worker_resources(
+            workers.user_owned_workers(self._model.root_user_id), running_bundles_info
+        )
```

Comment:
I think the only issue here is that the function `_deduct_worker_resources` happens before the real dispatching logic. Then when it calculates the resources left on user_owned_worker or public_worker, it always operates on the same set of `run_uuids`. With more and more bundles successfully getting dispatched, this may lead to task over-assigning issue, as `run_uuids` now has more values than before.

Reply:
oops you're totally right—this needs to be updated in the loop as well, but only to deduct more resources that are taken.

Reply:
i pushed a rather manual fix to this issue. There's some code duplication between here and `_deduct_bundle_resources`, not sure if we can do anything cleaner.

Reply:
Thanks very much! I think this will work. 
In terms of refactoring (it doesn't have to be done in this PR), I was wondering if we can do the following:
1. fetch users' parallel_run_quota once and store them in the dictionary `user_parallel_run_quota_left` that maps owner_id to number of parallel_run_quota left
1. dispatch bundles as usual except that we use `user_parallel_run_quota_left[owner_id]` now instead of querying the database 
1. subtract `user_parallel_run_quota_left[owner_id]` by 1 for each successfully dispatched bundle.

I think this way we can avoid resource deduction in two places.